### PR TITLE
:seedling: Add Refresh Session endpoint

### PR DESCRIPTION
**Pull request by AlainODea**
- Deprecate Extend Session endpoint (extendSession method)

 - Add Refresh Session endpoint (refreshSession method)

### CODE REVIEW

**Comment by bdemers** [comment]
Add a test?

**Comment by AlainODea** [comment]
@bdemers I'm on it.

I have another PR inbound for the Identity Providers API.

Any high-level suggestions on test strategy?

**Comment by bdemers** [comment]
@AlainODea I was just looking at our current tests, and we don't really have a set UT strategy. (there are a bunch of component/ITs though)

If I were going to suggest anything I would say Mockito and a spy.

All that said, I don't want to suggest doing a bunch of work either, as we are moving to the 'new-deal' branch soon.  This only covers the CRUD operations of the API  at the moment though.
We don't have much in the way of doc yet, and there is still a little cleanup to be done, but you can take an early look at it if you want: https://github.com/okta/okta-sdk-java/blob/new-deal/examples/quickstart/src/main/java/quickstart/Quickstart.java

**Comment by AlainODea** [comment]
No worries at all. Tests are important for maintaining a working implementation.

Mockito is good. I've used it before and I could assert things like correct JSON sent, correct URLs built, and correct deserialization of pre-canned API responses. I'll give that a shot.

**Comment by AlainODea** [comment]
@bdemers I'm lost here.

Mockito is not available and the code is not built for dependency inversion to facilitate that style of testing.

Stepping back and looking at the existing tests on master I tried to add a test in mock-okta for this and I can't get it to work.

I don't understand your unit test framework for this and cannot get it to work.

I created **src/test/java/SessionApiClientTest.java**:
```java
import com.okta.sdk.clients.SessionApiClient;
import com.okta.sdk.framework.ApiClientConfiguration;
import com.okta.sdk.models.sessions.Session;
import org.testng.annotations.Test;

import java.util.HashMap;
import java.util.Map;

public class SessionApiClientTest {

    private SessionApiClient getClient(String description) throws Exception {
        Map<String, String> customHeaders = new HashMap<>();
        customHeaders.put("x-test-description", description);
        TestConfig.MockOkta testConfig = Util.parseTestConfig().getMockOkta();
        return new SessionApiClient(
                new ApiClientConfiguration(
                        String.format("%s:%d", testConfig.getProxy(), testConfig.getPort()),
                        testConfig.getApiKey(),
                        customHeaders
                )
        );
    }

    @Test
    public void testRefreshSession() throws Exception {
        SessionApiClient sessionApiClient = getClient("/api/v1/sessions/:id/lifecycle/refresh - refreshes a session");
        Session session = sessionApiClient.refreshSession("1234fake");
    }
}
```

I created **tools/mock-okta/test/suites/sessions.js**:
```javascript
const util = require('../util');

util.describe('/api/v1/sessions/:id', (suite) => {
  suite.it('refreshes a session', req =>
    req.get({
      path: '/api/v1/sessions/1234fake/lifecycle/refresh',
      body: {
        "id": "101W_juydrDRByB7fUdRyE2JQ",
        "login": "user@example.com",
        "userId": "00ubgaSARVOQDIOXMORI",
        "expiresAt": "2015-08-30T18:41:35.818Z",
        "status": "ACTIVE",
        "lastPasswordVerification": "2015-08-30T18:41:35.818Z",
        "lastFactorVerification": null,
        "amr": [
          "pwd"
        ],
        "idp": {
          "id": "00oi5cpnylv792IcF0g3",
          "type": "OKTA"
        },
        "mfaActive": false,
        "_links": {
          "self": {
            "href": "https://your-domain.okta.com/api/v1/sessions/101W_juydrDRByB7fUdRyE2JQ",
            "hints": {
              "allow": [
                "GET",
                "DELETE"
              ]
            }
          },
          "refresh": {
            "href": "https://your-domain.okta.com/api/v1/sessions/101W_juydrDRByB7fUdRyE2JQ/lifecycle/refresh",
            "hints": {
              "allow": [
                "POST"
              ]
            }
          },
          "user": {
            "name": "Isaac Brock",
            "href": "https://your-domain.okta.com/api/v1/users/00uit00ZK6ELuzPoD0g3",
            "hints": {
              "allow": [
                "GET"
              ]
            }
          }
        }
      },
    })
  );
});
```

I launch mock-okta with npm:

```bash
npm run pretest
```

Then I run the SessionApiClient in IntelliJ and get this failure:

```plain
com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "req" (class com.okta.sdk.framework.ErrorResponse), not marked as ignorable (5 known properties: "errorLink", "errorCauses", "errorId", "errorCode", "errorSummary"])
 at [Source: org.apache.http.conn.EofSensorInputStream@6179e425; line: 2, column: 11] (through reference chain: com.okta.sdk.framework.ErrorResponse["req"])
```

When I debug I get a 404 status code back from the mock-okta Node.js server.

This is the full session output from running the test:

```plain
/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/bin/java -agentlib:jdwp=transport=dt_socket,address=127.0.0.1:54770,suspend=y,server=n -ea -Didea.test.cyclic.buffer.size=1048576 -Dfile.encoding=UTF-8 -classpath "/Users/alain.odea/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/171.4249.39/IntelliJ IDEA.app/Contents/lib/idea_rt.jar:/Users/alain.odea/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/171.4249.39/IntelliJ IDEA.app/Contents/plugins/testng/lib/testng-plugin.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/charsets.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/deploy.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/ext/cldrdata.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/ext/dnsns.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/ext/jaccess.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/ext/jfxrt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/ext/localedata.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/ext/nashorn.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/ext/sunec.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/ext/sunjce_provider.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/ext/sunpkcs11.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/ext/zipfs.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/javaws.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/jce.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/jfr.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/jfxswt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/jsse.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/management-agent.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/plugin.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/resources.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/jre/lib/rt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/lib/ant-javafx.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/lib/dt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/lib/javafx-mx.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/lib/jconsole.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/lib/packager.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/lib/sa-jdi.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_131.jdk/Contents/Home/lib/tools.jar:/Users/alain.odea/Documents/learning/java/oktasdk-java/target/test-classes:/Users/alain.odea/Documents/learning/java/oktasdk-java/target/classes:/Users/alain.odea/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.5.0/jackson-core-2.5.0.jar:/Users/alain.odea/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.5.0/jackson-databind-2.5.0.jar:/Users/alain.odea/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.5.0/jackson-annotations-2.5.0.jar:/Users/alain.odea/.m2/repository/joda-time/joda-time/2.7/joda-time-2.7.jar:/Users/alain.odea/.m2/repository/com/fasterxml/jackson/datatype/jackson-datatype-joda/2.5.0/jackson-datatype-joda-2.5.0.jar:/Users/alain.odea/.m2/repository/org/slf4j/slf4j-api/1.7.10/slf4j-api-1.7.10.jar:/Users/alain.odea/.m2/repository/org/apache/httpcomponents/httpclient/4.5.2/httpclient-4.5.2.jar:/Users/alain.odea/.m2/repository/org/apache/httpcomponents/httpcore/4.4.4/httpcore-4.4.4.jar:/Users/alain.odea/.m2/repository/commons-logging/commons-logging/1.2/commons-logging-1.2.jar:/Users/alain.odea/.m2/repository/commons-codec/commons-codec/1.9/commons-codec-1.9.jar:/Users/alain.odea/.m2/repository/com/google/guava/guava/18.0/guava-18.0.jar:/Users/alain.odea/.m2/repository/org/testng/testng/6.8.17/testng-6.8.17.jar:/Users/alain.odea/.m2/repository/org/beanshell/bsh/2.0b4/bsh-2.0b4.jar:/Users/alain.odea/.m2/repository/com/beust/jcommander/1.27/jcommander-1.27.jar:/Users/alain.odea/.m2/repository/org/hamcrest/hamcrest-core/1.3/hamcrest-core-1.3.jar:/Users/alain.odea/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/171.4249.39/IntelliJ IDEA.app/Contents/plugins/testng/lib/jcommander.jar" org.testng.RemoteTestNGStarter -usedefaultlisteners false -socket54769 @w@/private/var/folders/lk/kqwb1zxj11q29xn095vvygt4f__zjc/T/idea_working_dirs_testng.tmp -temp /private/var/folders/lk/kqwb1zxj11q29xn095vvygt4f__zjc/T/idea_testng.tmp
Connected to the target VM, address: '127.0.0.1:54770', transport: 'socket'
[TestNG] Running:
  /Users/alain.odea/Library/Caches/IntelliJIdea2017.1/temp-testng-customsuite.xml
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.

com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "req" (class com.okta.sdk.framework.ErrorResponse), not marked as ignorable (5 known properties: "errorLink", "errorCauses", "errorId", "errorCode", "errorSummary"])
 at [Source: org.apache.http.conn.EofSensorInputStream@6179e425; line: 2, column: 11] (through reference chain: com.okta.sdk.framework.ErrorResponse["req"])

	at com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException.from(UnrecognizedPropertyException.java:51)
	at com.fasterxml.jackson.databind.DeserializationContext.reportUnknownProperty(DeserializationContext.java:817)
	at com.fasterxml.jackson.databind.deser.std.StdDeserializer.handleUnknownProperty(StdDeserializer.java:954)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownProperty(BeanDeserializerBase.java:1315)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownVanilla(BeanDeserializerBase.java:1293)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:249)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:136)
	at com.fasterxml.jackson.databind.ObjectMapper._readValue(ObjectMapper.java:3532)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:1892)
	at com.fasterxml.jackson.core.JsonParser.readValueAs(JsonParser.java:1434)
	at com.okta.sdk.framework.JsonApiClient.unmarshall(JsonApiClient.java:69)
	at com.okta.sdk.framework.ApiClient.extractError(ApiClient.java:565)
	at com.okta.sdk.framework.ApiClient.checkResponse(ApiClient.java:550)
	at com.okta.sdk.framework.ApiClient.unmarshallResponse(ApiClient.java:516)
	at com.okta.sdk.framework.ApiClient.post(ApiClient.java:196)
	at com.okta.sdk.clients.SessionApiClient.refreshSession(SessionApiClient.java:173)
	at SessionApiClientTest.testRefreshSession(SessionApiClientTest.java:27)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:348)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:343)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:305)
	at org.testng.SuiteRunner.run(SuiteRunner.java:254)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:127)


===============================================
Default Suite
Total tests run: 1, Failures: 1, Skips: 0
===============================================

Disconnected from the target VM, address: '127.0.0.1:54770', transport: 'socket'

Process finished with exit code 0
```

**Comment by bdemers** [comment]
@AlainODea Thanks for looking into this!

@jmelberg-okta any ideas on this?

**Comment by jmelberg-okta** [comment]
@AlainODea: Thanks for the PR! 

We started building out a test suite that serves out pre-recorded responses to specific API calls. As you noticed, we generate these requests in `tools/mock-okta/test/suites/*.js`, and record the response in the form of a tape inside of `tools/mock-okta/tapes/`. All of the suites and tapes are served from a parent repository, so they can be distributed across all of our SDKs.

Here is where things get a bit tricky. We specify our `mockOkta` server inside of the configuration file`.sdk.config.json`, which contains both a proxied organization URL and an `apiKey`. When the mock-server is run, it references the values in this file for testing purposes. The error we see is most likely due to fact that there are no existing **session** tapes. Once tapes are created, the mock-server will know what response to return when when we hit the `testRefreshSession()` test. You can generate tapes from inside of this repo when running the command:
```sh
[okta-sdk-java] $ npm run record-new-tapes
```
**tl;dr**: My guess is that the errors we are seeing are due to not having session tapes. To test this locally, you can update the `.sdk.config.json` file to your organization and run the **record-new-tapes** script.

**Comment by AlainODea** [comment]
@jmelberg-okta great clarification. Thank you! I'll give this a shot next week.

**Comment by AlainODea** [comment]
@jmelberg-okta @bdemers I got it to generate new tapes without huge trouble, but **src/test/java/SessionApiClientTest.java** still fails.

Getting a valid session ID to try with the refresh endpoint requires creating a user, then using the [Authentication API with an apiToken to get a sessionToken](https://developer.okta.com/docs/api/resources/authn.html#primary-authentication-with-activation-token) to authenticate as that user, then using that sessionToken to obtain a usable session ID, then doing a refresh on that ID.

This suite for tape generation **test/mock-okta/test/suites/sessions.js** does those steps and appears to work:

```javascript
const util = require('../util');

util.describe('/api/v1/sessions/:id', (suite) => {
  const password = 'tlpWENT2m';
  suite.it('refreshes a session', req =>
    req.post({
      path: '/api/v1/users?activate=true',
      body: {
        profile: {
          firstName: 'Sessions',
          lastName: 'McJanky',
          email: 'mocktestexample-sessions@mocktestexample.com',
          login: 'mocktestexample-sessions@mocktestexample.com',
          mobilePhone: '555-415-1337',
        },
        credentials: {
          password: {
            value: password,
          },
        },
      },
    })
    .then(user =>
      req.post({
        path: '/api/v1/authn',
        body: {
          username: user.profile.login,
          password: password,
        },
      })
    )
    .then(authn =>
      req.post({
        path: '/api/v1/sessions',
        body: {
          sessionToken: authn.sessionToken,
        }
      })
    )
    .then(session =>
      req.post({
        path: `/api/v1/sessions/${session.id}/lifecycle/refresh`
      })
    )
  );
});
```

I am getting 404 Not Found when I debug **SessionApiClientTest**.

When I run it I get a JSON deserialization error:

```plain
com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "req" (class com.okta.sdk.framework.ErrorResponse), not marked as ignorable (5 known properties: "errorLink", "errorCauses", "errorId", "errorCode", "errorSummary"])
 at [Source: org.apache.http.conn.EofSensorInputStream@2fc6f97f; line: 2, column: 11] (through reference chain: com.okta.sdk.framework.ErrorResponse["req"])
```

I don't understand how mock-okta works as the expected tapes don't play back when **SessionApiClientTest** runs. I've written it to follow the precise set of API calls made in the tape generation suite.

I've pushed what I have. Can you folks take it from here?

**Comment by AlainODea** [comment]
I'm not a tire-kicker on PRs for what it's worth: https://github.com/okta/okta-signin-widget/pull/187

I will go the distance if you folks can prod me along occasionally :)

**Comment by bdemers** [comment]
@jmelberg-okta ^

**Comment by bdemers** [comment]
@jmelberg-okta ping ;)

**Comment by bdemers** [comment]
We are working this for an upcoming release (currently defining the swagger spec)

**Comment by AlainODea** [comment]
That works for me.
